### DISCUSSION TITLE
Beispielskript "Mitteilung: "Waschmaschine ist fertig"

### DISCUSSION THREAD

**Discussion by Mate883** [brutella/hkknx-public#334] [Q&A] [answered]
Hallo zusammen,

ich bin kein Coder, kenne mich daher im Detail nicht so gut aus. Ich habe versucht das Beispielskript zu implementieren, scheint jedoch leider nicht zu funktionieren. 

https://hochgatterer.me/de/hkknx/scripts/examples/washing-maschine-notification/

Ich habe die Gruppenadresse angepasst, auf dem Bus werden auch regelmäßig Werte vom Typ "7012 mA" gesendet. Aber das Skript schickt willkürlich die Meldungen. Passt evtl. der Part auch nicht "current = hkknx.ParseDPT9001(value)" ? Müsste es in dem Fall nicht "ParseDPT7012" sein? 

Danke

**Comment by brutella** [comment]
Ja, der Datentyp muss natürlich passen, ansonsten werden die Daten falsch dekodiert.

Testweise kannst du den dekodierten Wert in die Console schreiben und überprüfen welcher Wert dekodiert wurde.

```go
current = hkknx.ParseDPT9001(value)
println(current) // Gibt den Wert der Variable in der Console aus
```

**Reply by Mate883** [reply]
Danke für die Info....habe es angepasst, er gibt bei ausgeschaltetem Trockner den Wert "0" aus ... das sollte ja passen. Aber irgendwie scheint was noch nicht zu 100% zu funktionieren

```go
hkknx, time = import("hkknx"), import("time")

ch = hkknx.GroupWriteNotify("10/0/3")

// valueWithinRange returns true, if the value received from 10/0/3 
// stays within a range (min and max) for a specific duration.
valueWithinRange = func(min, max, duration) {
    timeout = time.After(duration)
    for {
        select {
		case value = <-ch:
            current = hkknx.ParseDPT7012(value)
			println(current)
            if current < min || current > max {
                return false
            }
        case <-timeout:
            return true
        }
    }
}

for {
    // Check if the electric current stays within 
    // 1 and 3 A for at least 10 minutes.
    if !valueWithinRange(1000, 3000, 10 * time.Minute) {
        continue
    }
    
    for {
        // Check if the electric current stays within 
        // 0 and 750 mA for at least 2 minute.
        if !valueWithinRange(0, 750, 2 * time.Minute) {
            continue
        }
        hkknx.SendNotification("Trocker fertig", "Bitte Leer machen.", "iPhone von XXX ")
        break // Break out of loop
    }
}
```

**Reply by brutella** [reply]
Die Stromaufnahme eines Trockners kann ganz anders sein, als wie bei einer Waschmaschine. Bevor du ein Skript erstellst, solltest du das Verhalten deines Trockners während einer Trocknung überwachen und dann entscheiden, welche Kriterien ausschlaggebend sind, dass der Trockner fertig ist.

Wenn du in hkknx die Aufzeichnung von Monitoring-Daten aktivierst hast, kannst du über die Web-Suche mit dem Suchtext "ga:10/0/3" den Verlauf der Stromaufzeichnung gut überwachen.

**Reply by Mate883** [reply]
Ja, die Werte habe ich eigentlich schon angepasst. Aber gut, dann muss ich das evtl. nochmal anschauen. Sofern kein Fehler in der Syntax drin ist, muss es ja woanders liegen

**Reply by brutella** [reply]
Du kannst auch gerne den Verlauf der Stromwerte hier posten, und wir erarbeiten gemeinsam an einer Lösung.

**Reply by Mate883** [reply]
Gerne :-) Anbei mal ein Screenshot aus Grafana am Beispiel Trockner, habe dasselbe auch für die Waschmaschine, aber wenn wir schon beim Trockner sind bleiben wir mal da: 


![000492](https://github.com/user-attachments/assets/5174ae5b-31d6-4fa1-a19a-5e72ff6a2801)

**Reply by Mate883** [reply]
Ich habe jetzt noch etwas rumprobiert...ohne Waschmaschine sondern mit einer einfachen Lampe die eingesteckt ist, um zu prüfen ob das Skript funktioniert. Leider klappt das auch nicht....er ignoriert die Duration und geht direkt nach einstecken in den nächsten Step und meldet dann auch direkt "Fertig" .... leider fehlt mir die Kenntnis da tiefer in die Fehleranalyse zu gehen :-(

**Reply by brutella** [reply, marked as answer]
So, ich denke, ich habe jetzt eine gute Lösung für das Problem gefunden. In Version [3.0.0-b9.2](https://github.com/brutella/hkknx-public/releases/tag/3.0.0-b9.2) kann man ganz einfach eine Überwachungslogik programmieren. 

Dazu erstellt man eine `ValueObservation` mit der Funktion `ObserveValues` für eine oder mehrere Gruppenadressen. (In deinem Fall wäre das eine Gruppenadresse – nehmen wir mal die `1/2/3`.)

Laut deinem Verlaufsdiagramm ist die Waschmaschine 
- aktiv, wenn 500 mA überschritten werden
- inaktiv, wenn 500 mA unterschritten werden.

Anhand dieser Informationen erstellen folgendes Skript.

```go
hkknx, time, log = import("hkknx"), import("time"), import("log")

// Überwache die Gruppenadresse 1/2/3
observe = hkknx.ObserveValues("1/2/3")

for {
	// Warte bis der Wert 500 mA für 5 Minuten überschritten wird.
	// Mit der Toleranz von 1 Minute werden Werte weniger als 500mA für 1 Minute toleriert.
	<-observe.GreaterThan(500).Duration(5 * time.Minute).Tolerance(1 * time.Minute).NotifyOnce()
	hkknx.SendNotification("Waschmaschine", "Die Waschmaschine wurde gestartet.")

	// Warte bis der Wert 500 mA für 5 Minuten unterschritten wird.
	<-observe.LessThan(500).Duration(5 * time.Minute).NotifyOnce()
	hkknx.SendNotification("Waschmaschine", "Die Waschmaschine ist fertig.")
}
```

Du kannst die Schwellenwerte und Dauer noch anpassen.

**Reply by Mate883** [reply]
Cool, danke! Das werde ich gleich mal ausprobieren

**Reply by brutella** [reply]
@Mate883 Hast du das Skript schon getestet?

**Reply by Mate883** [reply]
Aktuell noch nicht. Ich habe einen Workaround gebaut mittels KNX Bordmitteln. -> Zwei Schalter hinzugefügt (Waschmaschine und Trockner). Gesteuert werden die vom KNX Aktor. Wenn die Maschine läuft gehen die an, bei ausschalten bekomme ich eine Nachricht. Auf dem Aktor überwache ich die Werte. Allerdings gestaltet sich das bei meiner Waschmaschine schwierig......die zieht im Standby manchmal 4-5 Watt, geht aber während dem laufenden Waschprogramm auch mal darunter....sobald ich die Werte einigermaßen eingrenzen kann, teste ich dasselbe mal mit dem neuen Funktion

**Reply by brutella** [reply]
> Allerdings gestaltet sich das bei meiner Waschmaschine schwierig......die zieht im Standby manchmal 4-5 Watt, geht aber während dem laufenden Waschprogramm auch mal darunter

Wie im Skript bekommst du mit der Funktion [Tolerance](https://hochgatterer.me/hkknx/scripts/pkgs/hkknx/#ValueObservation.Tolerance) dieses Problem in den Griff. Damit werden Werte ausserhalb eines Grenzwertes geduldet.

**Reply by Mate883** [reply]
Habe die Skripte gerade mal eingebaut und teste jetzt mal....Danke!

**Reply by brutella** [reply]
Konntest du das Skript schon testen?

**Reply by Mate883** [reply]
Ja, es klappt soweit. Wobei meine Waschmaschine vom Verlauf her echt schwierig ist vernünftig zu tracken. Beim Trockner funktioniert es aber sehr gut. Danke nochmal
